Clarify metric attributes should be namespaced (#786)
open-telemetry/semantic-conventions: docs/general/metrics.md

the name does not appear as a key of any "rename_metrics" section (keys denote
old metric names in rename operations).

### Metric attributes

Metric attributes SHOULD follow the general [attribute naming rules](attribute-naming.md).
In particular, metric attributes SHOULD have a namespace.

Metric attributes SHOULD be added under the metric namespace when their usage and
semantics are exclusive to the metric.

Examples:

Attributes `mode` and `mountpoint` for metric `system.filesystem.usage`
should be namespaced as `system.filesystem.mode` and `system.filesystem.mountpoint`.

> Metrics can also have attributes outside of its namespace.

Reviewer: Maybe?
```suggestion
Metrics can also have attributes outside of their namespace.
```

the name does not appear as a key of any "rename_metrics" section (keys denote
old metric names in rename operations).

### Metric attributes

Metric attributes SHOULD follow the general [attribute naming rules](attribute-naming.md).

> In particular, metric attributes MUST have a namespace.

Reviewer 2: We discussed previously that MUST is stronger that in the naming doc 

> All names that are part of OpenTelemetry semantic conventions SHOULD be part of a namespace.

I'd still prefer to keep it consistent

```suggestion
In particular, metric attributes SHOULD have a namespace.
```

Author: Ah I thought I have changed that. Sure, that makes sense.

Metric attributes SHOULD be added under the metric namespace when their usage and
semantics are exclusive to the metric.

Examples:

Attributes `mode` and `mountpoint` for metric `system.filesystem.usage`
should be namespaced as `system.filesystem.mode` and `system.filesystem.mountpoint`.

Metrics can also have attributes outside the metric namespace when these come

> from the attributes registry.

Reviewer 3: "when these come from the attributes registry" sort of makes it sounds like otherwise they don't need to be in the attributes registry

Author: Yeah I don't love it as well.. but I thought it would be good to point that there could be metrics like that and to remove the possible confusion of "why these are not namespaced under the metric". Do you have any suggestion? 

I removed that portion, now it's just:

> Metrics can also have attributes outside of its namespace.

the name does not appear as a key of any "rename_metrics" section (keys denote
old metric names in rename operations).

### Metric attributes

> Metric attributes MUST follow the general [attribute naming rules](attribute-naming.md).

Reviewer 2: this is stronger than recommendation (SHOULD) in https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/attribute-naming.md#recommendations-for-opentelemetry-authors

Since attributes are not signal-specific, I think we should stay consistent.

Author: Good point. Adapted to SHOULD. Thanks!

In particular, metric attributes MUST have a namespace.

Metric attributes SHOULD be added under the metric namespace when their usage and
semantics are exclusive to the metric.

Examples:

> Metric `system.filesystem.usage` with attributes `mode` and `mountpoint`
> should be namespace as: `system.filesystem.mode` and `system.filesystem.mountpoint`.

Reviewer 3: (just a rewording suggestion)
```suggestion
Attributes `mode` and `mountpoint` for metric `system.filesystem.usage` should be namespaced as `system.filesystem.mode` and `system.filesystem.mountpoint`.
```

When metric attributes apply to other signals, these do not need

> to be namespace under the metric and SHOULD be added to the

Reviewer: ```suggestion
to be namespaced under the metric and SHOULD be added to the
```

> [attributes registry](../attributes-registry/README.md).

Reviewer 4: > SHOULD be added to the attribute registry

All attributes should be added to the registry, whether or not they are under the metric namespace, right?

Reviewer 2: yes, I'd be in favor of removing this paragraph completely and adding one to https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/attribute-naming.md#recommendations-for-opentelemetry-authors to recommend all attributes to be in the registry.

Author: I was a bit torn by this. If we add an attribute that is highly specific to a metric and there's no usage elsewhere, why would we want that to be on the registry? 🤔

Reviewer 2: I think we wanted everything to be in the registry, but that's not normative

> Attributes should be included in this registry before they are used in semantic conventions.

https://github.com/open-telemetry/semantic-conventions/blob/58c2705b7c93ea529d11db991e7ae4ad175473e1/docs/attributes-registry/README.md?plain=1#L15

From the yaml and codegen perspective there is no difference between attributes added to the registry and not added to one - the concept of registry is only captured in the folder structure currently, so I'm taking my comment back (as out of scope of this PR).

Reviewer: From my perspective, `discoverability` is one reason and specifically for future contributors. This means that if any future possible usages show up then we won't have to handle duplicates or the need to move to the registry then.

Author: Yes, discoverability is the only thing that comes to mind. But again, if it's an attribute that is very particular/specific to a single metric, would it still make sense to be in the registry? Given that maybe we will have a metric registry at some point.. I wonder..

I will add this to the meeting agenda today to see what others think about it.

Author: Ok discussed in the semconv meeting on 2024-03-04 and we DO want all attributes in the registry, regardless of their use in a single metric. I will change the wording here.

Author: @jack-berg @lmolkova I changed the wording, could you please take a look? Thanks! [e8c31c0](https://github.com/open-telemetry/semantic-conventions/pull/786/commits/e8c31c011c0a4c90651d86494b2b30f5f4b60d10)

Examples:

Metric `http.server.request.duration` uses attributes from the registry such as

> `server.port`, `error.type`

Reviewer 3: ```suggestion
`server.port`, `error.type`.
```
